python_modules/dagster/dagster/_daemon/asset_daemon.py

```diff
@@ -51,6 +77,72 @@ def get_current_evaluation_id(instance: DagsterInstance) -> Optional[int]:
     return AssetDaemonCursor.get_evaluation_id_from_serialized(raw_cursor) if raw_cursor else None
 
 
+class AutoMaterializeLaunchContext:
+    def __init__(
+        self,
+        tick: InstigatorTick,
+        instance: DagsterInstance,
+        logger: logging.Logger,
+        tick_retention_settings,
+    ):
+        self._tick = tick
+        self._logger = logger
+        self._instance = instance
+
+        self._purge_settings = defaultdict(set)
+        for status, day_offset in tick_retention_settings.items():
+            self._purge_settings[day_offset].add(status)
+
+    @property
+    def status(self) -> TickStatus:
+        return self._tick.status
+
+    @property
+    def logger(self) -> logging.Logger:
+        return self._logger
+
+    def add_run_info(self, run_id=None):
+        self._tick = self._tick.with_run_info(run_id)
+
+    def update_state(self, status: TickStatus, **kwargs: object):
+        self._tick = self._tick.with_status(status=status, **kwargs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(
+        self,
+        exception_type: Type[BaseException],
+        exception_value: Exception,
+        traceback: TracebackType,
+    ) -> None:
+        if exception_value and isinstance(exception_value, KeyboardInterrupt):
+            return
+
+        # Log the error if the failure wasn't an interrupt or the daemon generator stopping
+        if exception_value and not isinstance(exception_value, GeneratorExit):
+            error_data = serializable_error_info_from_exc_info(sys.exc_info())
+            tick_finish_timestamp = pendulum.now("UTC").timestamp()
+            self.update_state(
+                TickStatus.FAILURE, error=error_data, end_timestamp=tick_finish_timestamp
+            )
+
+        self._write()
+
+        for day_offset, statuses in self._purge_settings.items():
+            if day_offset <= 0:
+                continue
+            self._instance.purge_ticks(
```

Comment:
We should make sure we're getting the behavior we want in cloud.

Right now, `purge_ticks` is a no-op in cloud and we manage deletions at the cluster level.

Reply:
i'll double-check once its live, but it looks like the purge is purely based on timestamp in cloud, across all types: https://github.com/dagster-io/internal/blob/60cb4e0999a221a74aa16723f442bd7050c33c29/dagster-cloud/python_modules/dagster-cloud-backend/dagster_cloud_backend/daemon/cluster_daemon/cluster_gc_daemon.py#L30-L38